app/src/main/java/com/example/mykotlincalculator21/mvp/model/CalculatorModel.kt

```diff
@@ -1,52 +1,135 @@
+import com.example.mykotlincalculator21.utils.NumbersUtils.POSITION_ONE
+import com.example.mykotlincalculator21.utils.NumbersUtils.POSITION_ZERO
+import com.example.mykotlincalculator21.utils.NumbersUtils.ZERO_NUMBER_DOUBLE_TYPE
+import com.example.mykotlincalculator21.utils.OperandUtils
+import com.example.mykotlincalculator21.utils.StringUtils.EMPTY_STRING
+import com.example.mykotlincalculator21.utils.StringUtils.OPERATOR_DIVIDE
+import com.example.mykotlincalculator21.utils.StringUtils.OPERATOR_MINUS
+import com.example.mykotlincalculator21.utils.StringUtils.OPERATOR_MULTIPLY
+import com.example.mykotlincalculator21.utils.StringUtils.OPERATOR_SUM
 
 class CalculatorModel : CalculatorContract.CalculatorModelContract {
-    private var firstOperand: String = EMPTY_STRING
-    private var secondOperand: String = EMPTY_STRING
+    private val firstOperandUtils = OperandUtils()
+    private val secondOperandUtils = OperandUtils()
     private var operator: String = EMPTY_STRING
     private var result: String = EMPTY_STRING
+    private var error = ErrorUtils.NONE
 
     override fun saveNumber(number: String) {
         if (operator.isEmpty()) {
-            firstOperand += number
+            firstOperandUtils.addNumber(number)
         } else {
-            secondOperand += number
+            secondOperandUtils.addNumber(number)
         }
     }
 
-    override fun getPartialResult() = "$firstOperand$operator$secondOperand"
+    fun getValue(): String {
+        var firstValue = (firstOperandUtils.getValue()).toString()
+        var secondValue = (secondOperandUtils.getValue()).toString()
+        if (secondOperandUtils.isEmpty()) {
+            secondValue = EMPTY_STRING
+        }
+        return "$firstValue$operator$secondValue"
+    }
 
-    override fun getFullResult(): String {
-        var firstOperandValue = firstOperand.toDouble()
-        var secondOperandValue = secondOperand.toDouble()
+    override fun getPartialResult(): String {
+        return getValue()
+    }
 
-        return when (operator) {
-            OPERATOR_SUM -> (firstOperandValue + secondOperandValue).toString()
-            OPERATOR_MINUS -> (firstOperandValue - secondOperandValue).toString()
-            OPERATOR_DIVIDE -> (firstOperandValue / secondOperandValue).toString()
-            OPERATOR_MULTIPLY -> (firstOperandValue * secondOperandValue).toString()
-            else -> ERROR_MESSAGE
+    private fun isValidOperation(): Boolean {
+        if (operator.isEmpty()) {
+            if (firstOperandUtils.isEmpty()) {
+                result = EMPTY_STRING
+            } else {
+                result = java.lang.String.valueOf(firstOperandUtils.getValue())
+            }
+            return false
+        } else if (secondOperandUtils.isEmpty()) {
+            error = ErrorUtils.ERROR_MESSAGE_INVALID_FORMAT
+            result = EMPTY_STRING
+            return false
         }
+        return true
+    }
```

Comment:
in this case, you can use when without arguments, to avoid nested if-else

Reply:
Thx for the tip

Reply:
why this is a boolean funtion called validOperation and it sets also de Result?

when you call this function (line 61) you are setting again the result and the error. i think you can remove the writing of result here. right?